optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/uni/DroolsUniAccumulateFunctionBridge.java

```diff
@@ -18,29 +18,31 @@
 
 import java.io.ObjectInput;
 import java.io.ObjectOutput;
-import java.io.Serializable;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.WeakHashMap;
 import java.util.function.BiFunction;
 import java.util.function.Function;
 import java.util.function.Supplier;
 
 import org.kie.api.runtime.rule.AccumulateFunction;
 import org.optaplanner.core.api.score.stream.uni.UniConstraintCollector;
+import org.optaplanner.core.impl.score.stream.drools.common.DroolsAccumulateContext;
 
 /**
  * Drools {@link AccumulateFunction} that calls {@link UniConstraintCollector} underneath.
  * @param <A> input to accumulate
  * @param <ResultContainer_> implementation detail
  * @param <NewA> result of accumulation
  */
-final class DroolsUniAccumulateFunctionBridge<A, ResultContainer_ extends Serializable, NewA>
-        implements AccumulateFunction<ResultContainer_> {
+final class DroolsUniAccumulateFunctionBridge<A, ResultContainer_, NewA>
+        implements AccumulateFunction<DroolsAccumulateContext<ResultContainer_>> {
 
     private final Supplier<ResultContainer_> supplier;
     private final BiFunction<ResultContainer_, A, Runnable> accumulator;
     private final Function<ResultContainer_, NewA> finisher;
-    private final Map<Object, Runnable> undoMap = new HashMap<>(0);
+    // Weak in order to prevent a memory leak, as the the context is kept by Drools outside of our control.
+    private final Map<DroolsAccumulateContext<ResultContainer_>, Map<Object, Runnable>> undoMaps = new WeakHashMap<>(0);
```

Comment:
I needed to introduce a context carrier (`DroolsAccumulateContext`) with a stable `equals()`. (Therefore not the `ResultContainer_` itself.)

The reason why I didn't do this before was that until now, every accumulate was only operating on one set of data, and therefore there was really only one `context`. With the new, more complex rule, Drools started using different `context`s and the old code naturally broke.

Reply:
WeakHashMap might not do what you think it does (IIRC talks by either Goetz or Shipilev correctly) Let's discuss in bluejeans :)

Reply:
Put Map<Object, Runnable> into DroolsAccumulateContext, remove this Weak*Map.